solver/app/quote.go

```diff
@@ -1,63 +1,217 @@
+					Message: err.Error(),
+				},
+			})
+		}
+
+		var req QuoteRequest
+		if err := json.NewDecoder(rr.Body).Decode(&req); err != nil {
+			writeError(http.StatusBadRequest, errors.Wrap(err, "decode request"))
+			return
+		}
+
+		deposit, err := quoteFunc(ctx, req)
+		if r := new(RejectionError); errors.As(err, &r) { // RejectionError
+			write(QuoteResponse{
+				Rejected:          true,
+				RejectReason:      r.Reason.String(),
+				RejectDescription: r.Err.Error(),
+			})
+		} else if err != nil { // Error
+			writeError(http.StatusInternalServerError, err)
+		} else {
+			write(QuoteResponse{Deposit: &deposit}) // Success
+		}
+	})
+}
+
 // getQuote returns payment in `depositTkns` required to pay for `expenses`.
 //
 // For now, this is a simple quote that requires a single expense, paid
 // for by an equal amount of an equivalent deposit token. Token equivalence is
 // determined by symbol (ex arbitrum "ETH" is equivalent to optimism "ETH").
-func getQuote(depositTkns []Token, expenses []Payment) ([]Payment, RejectOrErr) {
+func getQuote(depositTkns []Token, expenses []Payment) ([]Payment, error) {
 	if len(depositTkns) != 1 {
-		return nil, RejectOrErr{
-			Reason: rejectInvalidDeposit,
-			Err:    errors.New("only single deposit token supported"),
-		}
+		return nil, newRejection(rejectInvalidDeposit, errors.New("only single deposit token supported"))
 	}
 
 	if len(expenses) != 1 {
-		return nil, RejectOrErr{
-			Reason: rejectInvalidExpense,
-			Err:    errors.New("only single expense supported"),
-		}
+		return nil, newRejection(rejectInvalidExpense, errors.New("only single expense supported"))
 	}
 
 	expense := expenses[0]
 	depositTkn := depositTkns[0]
 
 	if expense.Token.Symbol != depositTkn.Symbol {
-		return nil, RejectOrErr{
-			Reason: rejectInvalidDeposit,
-			Err:    errors.New("deposit token must match expense token"),
-		}
+		return nil, newRejection(rejectInvalidDeposit, errors.New("deposit token must match expense token"))
 	}
 
 	// make sure chain class (e.g. mainnet, testnet) matches
 	// we should reject with UnsupportedDestChain before this. the solver is
 	// initialized by network, which only includes chains of the same class
 	if expense.Token.ChainClass != depositTkn.ChainClass {
-		return nil, RejectOrErr{
-			Reason: rejectInvalidDeposit,
-			Err:    errors.New("deposit and expense must be of the same chain class (e.g. mainnet, testnet)"),
-		}
+		return nil, newRejection(rejectInvalidDeposit, errors.New("deposit and expense must be of the same chain class (e.g. mainnet, testnet)"))
 	}
 
 	return []Payment{
 		{
 			Token:  depositTkn,
 			Amount: expense.Amount,
 		},
-	}, RejectOrErr{}
+	}, nil
 }
 
 // coversQuote checks if `deposits` match or exceed a `quote` for expenses.
-func coversQuote(deposits, quote []Payment) RejectOrErr {
+func coversQuote(deposits, quote []Payment) error {
 	if len(quote) != len(deposits) {
-		return RejectOrErr{}
+		return errors.New("invalid params [BUG]")
```

Comment:
I assume this is unexpected and not actually valid.

Reply:
Actually can remove this

Reply:
or maybe i lied

Reply:
nvm unrelated test failed